opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -564,6 +561,81 @@ def _format_links(links):
         return f_links
 
 
+class _Limits:
+    """The limits that should be enforce on recorded data such as events, links, attributes etc.
+
+    This class does not enforce any limits itself. It only provides an a way read limits from env,
+    default values and in future from user provided arguments.
+
+    All limit must be a either non-negative integers or None.
+    Setting a limit to ``None`` will not set an limits for that data type.
+
+    Args:
+        max_events: Maximum number of events that can be added to a Span.
+        max_links: Maximum number of links that can be added to a Span.
+        max_attributes: Maximum number of attributes that can be added to a Span.
+    """
+
+    max_attributes: int
+    max_events: int
+    max_links: int
+
+    def __init__(
+        self,
+        max_events: Optional[int] = None,
+        max_links: Optional[int] = None,
+        max_attributes: Optional[int] = None,
+    ):
+        self.max_attributes = max_attributes
+        self.max_events = max_events
+        self.max_links = max_links
+
+    def __repr__(self):
+        return "max_attributes={}, max_events={}, max_links={}".format(
+            self.max_attributes, self.max_events, self.max_links
+        )
+
+    @classmethod
+    def _create(cls):
```

Comment:
Why not add this code in `__init__`?

Reply:
Because `None` is a valid value for limits which means no limit has been set. If we do this in __init__, there is not way to know if `None` was explicily passed by the user to not set a limit or if the value was not provided at all in which case it should be loaded from env with fallback on default value.


This is more of a remnant from the earlier version which allowed users to set limits via code. In that version I also had a constant called `limits.UNSET` which could be passed to not set a value but since we removed the ability to set limits via code, this is probably not needed anymore. We can do two things here for now:

1. move this logic to __init__ and add a new `_create_empty()` method that returns limits with all values unset. This will be used by the default "unset limits" instance that is used as the default value for spans. It feels a bit less flexible overall as it means either use `_Limits()` and always load default for missing kwarg or set all limits to unset by calling `_Limits._crate_empty()`.

2. second option is to add the `UNSET` constant back. That way we can merge `_create()` into `__init__` and allow setting any individual limit to UNSET as `_Limits(max_events=_Limits.UNSET, max_links=100)`. In this example, there will be no upper limit on events, links will be limits to 100 and any other limits will be loaded from env or default values. I like this the most from an API point of the view but since this is not supposed to be used by users, I tried to simplify it by removing it. If you prefer this option, I can add it back.

Reply:
Implemented option 2. and got rid of the `_create()` method.